### PR TITLE
Plugins that reserved must unreserve when another plugin fails to reserve

### PR DESCRIPTION
**Pull request by mrkm4ntr**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
If one of reserve plugins failed, the plugins already reserved will not be unreserved.
At least this will be a problem when migrating TAVP. #82934 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 1]
```diff
@@ -426,13 +433,22 @@ func (f *framework) RunPostBindPlugins(
 // returns the error. In such case, pod will not be scheduled.
 func (f *framework) RunReservePlugins(
 	state *CycleState, pod *v1.Pod, nodeName string) *Status {
+	reserved := []string{}
+	defer func() {
+		for _, name := range reserved {
+			if pl, ok := f.unreservePlugins[name]; ok {
```
Comment (hex108) [inline pkg/scheduler/framework/v1alpha1/framework.go]: It assumes that the reserved plugin also implements `Unreserve` plugin interface.  It is reasonable, but we do not make a promise about it now. How about adding a restriction that a `Reserve` plugin also need implements `Unreserve` plugin interface (in another PR)? cc @ahg-g @Huang-Wei  

If so, we do not need a map `f.unreservedPlugins` to record it.

**Review by hex108 — COMMENTED** [review]
Thanks @mrkm4ntr! 

Another thought: is it OK to call `fwk.RunUnreservePlugins` at https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go#L634 if failing to run reserved plugins? If some reserved plugins have not started to run in the func `fwk.RunUnreservePlugins`, it could skip unreserve operations by itself.

**Comment by mrkm4ntr** [comment]
@hex108 Thank you for your review.

> If some reserved plugins have not started to run in the func fwk.RunUnreservePlugins, it could skip unreserve operations by itself.

You mean those who implement plugins take care about that the unreserve extension point may be called without calling of the reserve extension point? If so, it's ok just to run `fwk.RunUnreservePlugins`.

File: pkg/scheduler/framework/v1alpha1/framework.go (hunk 1 above)
Comment (mrkm4ntr) [reply pkg/scheduler/framework/v1alpha1/framework.go]: > How about adding a restriction that a Reserve plugin also need implements Unreserve plugin interface (in another PR)?

Sounds good! It seems to be natural to make `Reserve` and `Unreserve` be pair.

**Comment by hex108** [comment]
> You mean those who implement plugins take care about that the unreserve extension point may be called without calling of the reserve extension point? If so, it's ok just to run fwk.RunUnreservePlugins.

Yes :)

**Comment by mrkm4ntr** [comment]
Umm... I think your idea (merging reserve and unreserve) is better than the implicit convention (unreserve without reserve). I will create an issue about merging reserve and unreserve plugins and apply the conclusion of the issue to this PR.

**Comment by hex108** [comment]
@mrkm4ntr I think it will be better to resolve #83788 first.

**Comment by draveness** [comment]
> @mrkm4ntr I think it will be better to resolve #83788 first.

+1

**Pull request closed without merging**